### PR TITLE
Added a warning message for infinite epochs

### LINKED ISSUE #11554: Warn when running an infinite epoch and overriding "epoch end" accumulating hooks
## 🚀 Feature

When the user configures `Trainer(max_steps=-1, max_epochs=-1)` an endless epoch runs, so overriding `training_epoch_end` or `validation_epoch_end` with `val_check_interval==float` can be a problem because they will keep outputs in memory indefinitely.

### Motivation

Many users are not aware of the impact of overriding these hooks so infinite epochs open the door to "memory leaks".

### Pitch

Raise a warning in this case informing the user of this behaviour.

### Additional context

Proposed in https://github.com/PyTorchLightning/pytorch-lightning/issues/11480#issuecomment-1014594779

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @carmocca @awaelchli @ninginthecloud @daniellepintz @rohitgr7 @justusschock @kaushikb11

### PR DESCRIPTION
**Pull request by VedPatwardhan**
## What does this PR do?

Fixes #11554 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by akihironitta — COMMENTED** [review]
_

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -429,6 +429,13 @@ def __init__(
         """
         super().__init__()
         Trainer._log_api_event("init")
+
+        # warning for infinite epochs
+        if max_steps == -1 and max_epochs == -1:
+            warnings.warn(
+                "This configuration can result in infinite epochs, as a result of overriding training_epoch_end or validation_epoch_end with val_check_interval==float. This will keep the outputs in memory indefinitely."
+            )
```
Comment (akihironitta) [inline pytorch_lightning/trainer/trainer.py:437]: Hi @VedPatwardhan! Thank you for working on this!

Writing it here leads to raising the warning to other users who didn't override the hooks, so I think we should check if the hooks are overridden by user, too. You can see other warnings defined in configure_validator.py:

https://github.com/PyTorchLightning/pytorch-lightning/blob/5ad5ba54c0c477546d21daf75ac7b4748d7963a7/pytorch_lightning/trainer/configuration_validator.py#L243

Comment (VedPatwardhan) [reply pytorch_lightning/trainer/trainer.py:437]: Got it. I'll make that change.

Comment (VedPatwardhan) [reply pytorch_lightning/trainer/trainer.py]: Could you please check it now?

Comment (VedPatwardhan) [reply pytorch_lightning/trainer/trainer.py]: @akihironitta could you please help me out understanding what's going wrong ?

File: pytorch_lightning/trainer/configuration_validator.py [hunk 2]
```diff
@@ -135,6 +137,17 @@ def __verify_train_val_loop_configuration(trainer: "pl.Trainer", model: "pl.Ligh
             " Please use `val_dataloader()` directly."
         )
 
+    # ----------------------------------------------
+    # verify if model leads to infinite epochs
+    # ----------------------------------------------
+    causes_infinite_epochs = is_overridden("training_epoch_end") or (
+        is_overridden("validation_epoch_end") and type(model.val_check_interval) == float
+    )
```
Comment (Borda) [inline pytorch_lightning/trainer/configuration_validator.py:145]:
```suggestion
    val_epoch_ = is_overridden("validation_epoch_end") and type(model.val_check_interval) == float
    causes_infinite_epochs = is_overridden("training_epoch_end") or val_epoch_
```

**Comment by carmocca** [comment]
Closing as this PR has become stale. We'd be looking to merge it if somebody wants to take over. Thank you for your time!

**Comment by VedPatwardhan** [comment]
Hi @carmocca, please take a look again, I think @Borda made the changes he had suggested. Thanks!

File: src/pytorch_lightning/trainer/configuration_validator.py [hunk 3]
```diff
@@ -117,6 +118,18 @@ def __verify_train_val_loop_configuration(trainer: "pl.Trainer", model: "pl.Ligh
...
+    # verify if model leads to infinite epochs
+    # ----------------------------------------------
+    val_epoch_ = is_overridden("validation_epoch_end") and type(model.val_check_interval) == float
+    causes_infinite_epochs = is_overridden("training_epoch_end") or val_epoch_
+    if causes_infinite_epochs:
+        rank_zero_warn(
+            "This configuration can result in infinite epochs, as a result of overriding `training_epoch_end`"
+            " or `validation_epoch_end` with `val_check_interval==float`."
+            " This will keep the outputs in memory indefinitely."
+        )
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/configuration_validator.py:131]: perhaps we should split it into two warnings, one for each override so it is clear to the user which one applies.
